lib/ecto/type.ex

```diff
@@ -533,23 +524,43 @@ defmodule Ecto.Type do
   defp load_time_usec(%Time{} = time), do: {:ok, pad_usec(time)}
   defp load_time_usec(_), do: :error
 
-  defp load_naive_datetime(%NaiveDateTime{} = naive_datetime), do: {:ok, truncate_usec(naive_datetime)}
+  # This is a downcast, which is always fine, and in case
+  # we try to send a naive datetime where a datetime is expected,
+  # the adapter will either explicitly error (Postgres) or it will
+  # accept the data (MySQL), which is fine as we always assume UTC
+  defp load_naive_datetime(%DateTime{} = datetime),
+    do: {:ok, datetime |> check_utc_timezone!(:naive_datetime) |> DateTime.to_naive() |> truncate_usec()}
+
+  defp load_naive_datetime(%NaiveDateTime{} = naive_datetime),
+    do: {:ok, truncate_usec(naive_datetime)}
+
   defp load_naive_datetime(_), do: :error
 
-  defp load_naive_datetime_usec(%NaiveDateTime{} = naive_datetime), do: {:ok, pad_usec(naive_datetime)}
+  defp load_naive_datetime_usec(%DateTime{} = datetime),
+    do: {:ok, datetime |> check_utc_timezone!(:naive_datetime_usec) |> DateTime.to_naive() |> pad_usec()}
+
+  defp load_naive_datetime_usec(%NaiveDateTime{} = naive_datetime),
+    do: {:ok, pad_usec(naive_datetime)}
+
   defp load_naive_datetime_usec(_), do: :error
 
-  defp load_utc_datetime(%DateTime{} = datetime),
-    do: {:ok, truncate_usec(datetime)}
+  # This is an upcast but because we assume the database
+  # is always in UTC, we can perform it.
   defp load_utc_datetime(%NaiveDateTime{} = naive_datetime),
-    do: naive_datetime |> truncate_usec() |> DateTime.from_naive("Etc/UTC")
+    do: {:ok, naive_datetime |> truncate_usec() |> DateTime.from_naive!("Etc/UTC")}
```

Comment:
Does this mean that NaiveDateTime now always loads as a DateTime in UTC time zone?

Reply:
No, this is for `utc_datetime`. The behaviour here didn't change. I just rewrote the code to be in sync with everything else. :)